Handled expiring complimentary subscriptions

rshbhgrg opened the pull request:
refs https://github.com/TryGhost/Team/issues/1727

- adds method to get active products for a member to exclude expired comped subs
- wires member add/edit to update expiry of a comped sub
- updates admin ui to allow adding expiry

naz commented on ghost/core/core/server/services/members/jobs/clean-expired-comped.js:
```diff
@@ -0,0 +1,71 @@
+    });
+}
+
+(async () => {
```
While we are at it, @rishabhgrg can you please check if we still need this self-execution function syntax. AFAIR our current range of Node versions supports top level await :thinking:

rishabhgrg replied:
From a quick check, top-level await seems to be still [only available in ES modules so we'll probably have to keep this style for now, but I'll come back to it later today to make sure that's the case.

naz commented on ghost/core/core/server/services/members/jobs/clean-expired-comped.js:
```diff
@@ -0,0 +1,71 @@
+    const db = require('../../../data/db');
+    debug(`Starting cleanup of expired comp subscriptions on ${cleanupStartDate}`);
+    const expiredCompedRows = await db.knex('members_products')
+        .where('expiry_at', '<', moment.utc().startOf('day').toISOString())
```
Is the logic here to expire products from yesterday and below or should include the ones set to expire at midnight (5 minutes before the run of this job)? in other words should it be `<` or `<=` :thinking:

rishabhgrg replied:
> Is the logic here to expire products from yesterday and below

It should be this, basically any sub whose expiry date is already in past and doesn't include today should be expired, as the ones that are expiring midnight means they still have access to tier till today.

naz commented on ghost/core/core/server/services/members/jobs/clean-expired-comped.js:
```diff
@@ -0,0 +1,71 @@
+        // Delete all expired comped rows
+        deletedExpiredSubs = await db.knex('members_products')
+            .whereIn('id', rowIds)
+            .del();
```
by removing these records my understanding is we don't want to know in the future if the member ever has been on the "expiring comped" plan?

rishabhgrg replied:
Yeah don't think we have any usecase to keep that info right now, but that did strike a good point to add member status event update as part of this job, so we do have a record of member's status update in the DB when they go from comped to free. 💡 Adding that now

rishabhgrg replied:
Done, added logic to insert member status events based on members going from comped -> free. We ofc still lose the expiry information, but don't think we need to keep that

naz replied:
oh this is a neat workaround!

naz commented on ghost/core/core/server/services/members/jobs/jobs.js:
```diff
@@ -0,0 +1,27 @@
+                name: 'clean-expired-comped'
+            });
+
+            hasScheduled = true;
```
The method only ever runs once in the Ghost process lifecycle - during the boot during `members.init()` execution. Why do we need this protective code from running the method multiple times? If it is run multiple times does it signal something wrong is going on?

rishabhgrg replied:
🤔 So i picked that pattern from email analytics, which I get has this check because it is run either on init or sometime down the line based on if emails need to be fetched. I am hoping comp check will also move in a similar direction maybe in the cleanup, so i kept the pattern intact as there is no reason to not have it imo. 
I don't think it will be run multiple times right now so i am ok either ways on keeping or removing this check, but also i don't think there is any harm in keeping it as well :)

naz commented on ghost/core/core/server/services/members/service.js:
```diff
@@ -6,6 +6,7 @@ const db = require('../../data/db');
 const MembersConfigProvider = require('./config');
 const MembersCSVImporter = require('@tryghost/members-importer');
 const MembersStats = require('./stats/members-stats');
+const memberJobs = require('./jobs/jobs');
```
I think leaving the pattern that exists in the `analytics` jobs (index.js file in the root) would be best for consistency sake. It should not be full of logic though :D

rishabhgrg replied:
I wanted to keep away from overloading `index.js` file as we generally avoid them, no strong prefs tho so happy to change it back to index here :)

rishabhgrg replied:
Updated back to `index.js` ✔️

naz commented on ghost/core/core/server/services/members/jobs/clean-expired-comped.js:
```diff
@@ -0,0 +1,72 @@
+    debug(`Removed ${deletedExpiredSubs} expired subscriptions, updated ${updatedMembers} members in ${cleanupEndDate.valueOf() - cleanupStartDate.valueOf()}ms`);
+
+    if (parentPort) {
+        parentPort.postMessage(`Removed ${deletedExpiredSubs} expired subscriptions, updated ${updatedMembers} members in ${cleanupEndDate.valueOf() - cleanupStartDate.valueOf()}ms`);
```
I'm thinking if we would end up with logs full of `Removed 0 expired subscriptions...` on majority of instances :thinking: It is probably fine in the beginning. How do you think we should handle it if that's the case?

rishabhgrg replied:
🤔 Yea that's a fair point, we could remove the logs for when nothing happens in future if that becomes annoying maybe and only output stuff when any changes are done ? Even with optimisations, don't think we can completely avoid not running jobs when there are no expiring subs, so there will always be some case where it runs but doesn't do anything

naz reviewed (approved):
Did look through job-related part of the change. Left a few clarification questions. For the most part looks :ok_hand: :ok_hand: :ok_hand:

rshbhgrg committed:
Added feature flag for cardless trials

refs https://github.com/TryGhost/Team/issues/1727

- adds alpha flag to toggle cardless trials via expiring complimentary subscriptions

rshbhgrg committed:
Handled storing complimentary subscription expiry

refs https://github.com/TryGhost/Team/issues/1727

- if feature flag is enabled, handles storing expiry date on complimentary subscriptions in `expiry_at` column of `members_products`
- updates the expiry value on both member edit or add with tiers
- expiry is passed as `expiry_at` in `tiers` list of a member
- includes `expiry_at` on tiers data of a member when flag is enabled

rshbhgrg committed:
Added scheduled job to clean expired complimentary subs

refs https://github.com/TryGhost/Team/issues/1727

- runs a daily cron job at start of the day to cleanup all expired comped subs
- removes `members<>products` mapping for expired entries, and updates status for corresponding members
- also adds status events for members going back from comp -> free as a result of expiry
- scope for future optimisation on how the scheduled job is ran or does the cleanup

rshbhgrg committed:
Wired expiring comp subscriptions on Admin

refs https://github.com/TryGhost/Team/issues/1727

- adds alpha feature flag for expiring comped subs
- adds duration to complimentary subs to set expiry
- shows details for expiring comped subs

ErisDS reviewed (approved):
If @naz is happy, I am happy :)
